### PR TITLE
new Datatype: Order

### PR DESCRIPTION
**Pull request by nhaagen**
A list of tuples (subject/direction) to express ordering
ORDEEER

### CODE REVIEW

**Comment by mjansenDatabay** [comment]
I highly appreciate this. I hate it to re-validate strings (passed from the UI layer into persistence layer, maybe crossing a few other layers) and use code like this: `if (!in_array(strtolower($orderDirection), [desc', 'asc'])) {$orderDirection = 'asc';}`.

File: src/Data/Factory.php [hunk 1]
```diff
@@ -163,4 +163,14 @@ public function range(int $start, int $length) : Range
     {
         return new Range($start, $length);
     }
+
+    /**
+     * @param string $id
+     * @param mixed $direction Order::ASC|Order::DESC
+     * @return Order
```
Comment (klees) [inline src/Data/Factory.php]:
```suggestion
     * @param mixed $direction Order::ASC|Order::DESC
```

[We MAY drop doc-blocks if they do not add any information.](https://github.com/ILIAS-eLearning/ILIAS/blob/release_6/src/README.md) Also, this very much clarifies how doc-blocks could be a problem: The block talks about `$id`, while the parameters use the name `$subject`. If we don't duplicate the typehint the problem vanishes as well =)

File: src/Data/Order.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        $clone->order[$subject] = $direction;
+        return $clone;
+    }
+
+    public function get() : array
+    {
+        return $this->order;
+    }
+
+    public function join(callable $fn, $prefix)
```
Comment (klees) [inline src/Data/Order.php]: Let's switch the two parameters. Conceptually, `$prefix` comes left to the function we are using and the function itself also uses that order in its parameters. Also, add a doc-block for this method. Finally, let's rename `$prefix` to `$init` (if you could not think of anything better) to make unambiguously clear that we indeed do not care if this creates a string or anything else.

**Review by klees — CHANGES_REQUESTED** [review]
Hi @nhaagen,

LGTM in general, just two minor nitpicks, then I'll be happy =)

Thx and best regards!

**Pull request merged**